### PR TITLE
[#9382] Migrate part of instructor students page to Angular

### PR DESCRIPTION
**Pull request by wenmogu**
Part of #9382 

Things to note:
  - BE:
    - Added some auxiliary methods to get data conveniently from typicalBundle for BaseActionTests. I know this is not the place. I will migrate the methods to a new class.
    - **Question/possibly bug: it always throws nullPointerException when I iterate typicalBundle.instructors with forEach or for-loop. I tried converting its values to stream and filter it, doesnt work as well. Such iteration and filtering works perfectly fine for other maps in typicalBundle.**
     - For getting the sections, teams, students of a course, everything is transmitted within one call in a single json object
  - FE:
    - Reuse component Student List Table
    - Student List Table: I made the table horizontally scrollable. If not, when the screen shrinks, the data display will become messy.
    - Students are identified by student email along, as there are no students using the same email address in one course.

Checklist:
- [x] BE
- [x] BE Tests
  - [x] BE Test 1
  - [x] BE Test 2
  - [ ] Data Bundle auxiliary methods migration
- [ ] BE deletion
- [ ] FE
  - [x] Checkboxes
    - [x] interactions among courses, sections, teams and students
    - [x] interactions among (select all)s and the items
    - [x] interaction between filtering and table display of students
  - [x] Search box
  - [ ] Table
    - [x] display students data
    - [x] hide students upon filtering
    - [ ] Sorting of columns
    - [x] Enroll button on the header
    - [ ] GIf when fetching data from BE
- [x] FE Test
- [ ] FE Deletion

Pic:
image

image

### CODE REVIEW

File: src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.html [hunk 1]
```diff
@@ -1,3 +1,153 @@
...
+<ng-container *ngIf="courses && courses.length" >
+  <div class="p-2">
+    <div *ngFor="let course of courses" class="card border mb-3 p-0"
+         [hidden]="(!course.isChecked) && (!isDisplayArchive) && course.isArchived">
+      <div class="card-header font-weight-bold"
+           [ngClass]="{'text-dark-blue bg-light-blue ': !course.isArchived, 'bg-light-grey': course.isArchived}"
+           (click)="toggleCourseStateAtInput(course)">
+        <strong>[{{course.id}}]:</strong> {{course.name}}
+        <div class="arrow-sign">
+          <a class="btn btn-info btn-xs" routerLink="/web/instructor/home" (click)="$event.preventDefault();"> Enroll </a>
```
Comment (wkurniawan07) [inline src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.html]: The routing for the enroll page should be fixed.

File: src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.scss [hunk 2]
```diff
@@ -0,0 +1,47 @@
+.bg-light-blue {
+  background-color: #007BFF33 !important;
```
Comment (wkurniawan07) [inline src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.scss]: Is `007BFF33` a valid hex color?

File: src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.scss [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+.border-light-blue {
+  border-color: #007BFF33 !important;
+}
+
+.text-dark-blue {
+  color: var(--blue) !important;
+}
+
+.fill-plain {
+  background-color: #EAEFF5 !important;
```
Comment (wkurniawan07) [inline src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.scss]: Are all the `!important`s above necessary?

File: src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts [hunk 4]
```diff
@@ -1,5 +1,62 @@
...
+import { StatusMessageService } from '../../../services/status-message.service';
+import { ErrorMessageOutput } from '../../error-message-output';
+import { StudentListSectionData, StudentListStudentData } from '../student-list/student-list-section-data';
+
+interface Course {
+  id: string;
+  name: string;
+  isArchived: boolean;
+  isInstructorAllowedToModify: boolean;
+  isChecked: boolean | undefined;
```
Comment (wkurniawan07) [inline src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts]: Why do all of these need to be `| undefined`?

File: src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts [hunk 5]
```diff
@@ -12,13 +69,490 @@ import { ActivatedRoute } from '@angular/router';
...
+  }
+
+  get isAllPresentCoursesChecked(): boolean {
+    if (this.allPresentCourses.length === 0) {
+      return false;
+    }
+    return this.allPresentCourses.length === this.allCheckedCourses.length;
+  }
+
+  get isAllPresentCourseDetailssChecked(): boolean {
```
Comment (wkurniawan07) [inline src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts]: Details~s~

File: src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts [hunk 6]
```diff
@@ -12,13 +69,490 @@ import { ActivatedRoute } from '@angular/router';
...
+
+  get allCheckedStudentDetails(): StudentDetails[] {
+    return this.allPresentStudentDetails.filter((studentDetails: StudentDetails) => studentDetails.isChecked);
+  }
+
+  get isAllPresentCoursesChecked(): boolean {
+    if (this.allPresentCourses.length === 0) {
+      return false;
+    }
+    return this.allPresentCourses.length === this.allCheckedCourses.length;
```
Comment (wkurniawan07) [inline src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts]: You can shorten this to
```ts
return this.allPresentCourses.length && this.allPresentCourses.length === this.allCheckedCourses.length;
```
Similarly for the rest below.

File: src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts [hunk 7]
```diff
@@ -12,13 +69,490 @@ import { ActivatedRoute } from '@angular/router';
...
+      }
+    });
+  }
+
+  /**
+   * Function to change states of all courses.
+   */
+  toggleAllPresentCoursesStateAtInput(defaultState: boolean): void {
+    this.allPresentCourses.forEach((course: Course) => {
+      if (course !== undefined) {
```
Comment (wkurniawan07) [inline src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts]: `if (course)` is sufficient?

File: src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts [hunk 8]
```diff
@@ -12,13 +69,490 @@ import { ActivatedRoute } from '@angular/router';
...
+   */
+  fetchCourseDetails(courseid: string): Observable<CourseDetails | undefined> {
+    if (this.isCourseDetailsInList(courseid)) {
+      return of(this.getCourseDetails(courseid));
+    }
+
+    const paramMap: { [courseid: string ]: string } = { courseid };
+    return this.httpRequestService.get('/instructor/students', paramMap)
+        .pipe(map((resp: GetCourseDetailsResponse | undefined) => {
+          if (resp !== undefined) {
```
Comment (wkurniawan07) [inline src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts]: Will it ever be `undefined`, though?

File: src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts [hunk 9]
```diff
@@ -12,13 +69,490 @@ import { ActivatedRoute } from '@angular/router';
...
+  }
+
+  get isAnyPresentCoursesChecked(): boolean {
+    if (this.allPresentCourses.length === 0) {
+      return false;
+    }
+    return this.allCheckedCourses.length > 0;
+  }
+
+  courseStudentListSectionDataMap: Map<string, StudentListSectionData[]> = new Map();
```
Comment (wkurniawan07) [inline src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts]: Use `{ [key: string]: StudentListSectionData[] }` as the type instead; it does not need to handle `| undefined` type.

File: src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts (hunk 4 above)
Comment (wenmogu) [reply src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts]: this is because this parameter is not passed from the backend, and yet i need it to be part of this object...

File: src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts (hunk 6 above)
Comment (wenmogu) [reply src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts]: I couldnt do this as `this.allPresentCourses.length` is not a boolean, which gives me compilation error

File: src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts (hunk 9 above)
Comment (wenmogu) [reply src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts]: I changed this based on the review, but i dont really get the part `it does not need to handle | undefined` type. If i use the object and I want to obtain a value given a key, I suppose I still need to check if this value exists in the object...? If this is the case, I prefer using Map, as it comes with a lot of other methods that make things more convenient.

**Comment by wenmogu** [comment]
@wkurniawan07 thx for the review! n so sorry abt my slowness ;_;
what abt this tho?

> * Question/possibly bug: it always throws nullPointerException when I iterate typicalBundle.instructors with forEach or for-loop. I tried converting its values to stream and filter it, doesnt work as well. Such iteration and filtering works perfectly fine for other maps in typicalBundle.

File: src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts (hunk 6 above)
Comment (wkurniawan07) [reply src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts]: Why would it give you compile-time error? JS/TS supports boolean-coercion.

Comment (wenmogu) [reply src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts]: I got something like this:

image

File: src/web/app/pages-instructor/student-list/student-list.component.html [hunk 10]
```diff
@@ -19,40 +23,41 @@
...
+    </thead>
+    <tbody>
+    <ng-container *ngFor="let section of sections">
+      <tr *ngFor="let student of section.students; trackBy: trackByFn" [hidden]="isStudentToHide(student.email)"s>
+        <td class="align-center">
+          <ng-template #popContent><img class="profile-pic" src="{{student.photoUrl}}"></ng-template>
+          <img *ngIf="student.photoUrl; else viewPhotoBtn" class="profile-pic-icon" src="{{student.photoUrl}}"
+               [ngbPopover]="popContent" triggers="mouseenter:mouseleave">
+          <ng-template #viewPhotoBtn>
+            <a href="#" role="button" (click)="loadPhoto(student); $event.preventDefault();">View Photo</a>
```
Comment (wkurniawan07) [inline src/web/app/pages-instructor/student-list/student-list.component.html]: Change `href="#"` to `[routerLink]=""`.

File: src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts (hunk 4 above)
Comment (wkurniawan07) [reply src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts]: Then you're looking for `isChecked?: boolean`.

File: src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.html [hunk 11]
```diff
@@ -1,3 +1,153 @@
...
+<ng-container *ngIf="courses && courses.length" >
+  <div class="p-2">
+    <div *ngFor="let course of courses" class="card border mb-3 p-0"
+         [hidden]="(!course.isChecked) && (!isDisplayArchive) && course.isArchived">
+      <div class="card-header font-weight-bold"
+           [ngClass]="{'text-dark-blue bg-light-blue ': !course.isArchived, 'bg-light-grey': course.isArchived}"
+           (click)="toggleCourseStateAtInput(course)">
+        <strong>[{{course.id}}]:</strong> {{course.name}}
+        <div class="arrow-sign">
+          <button class="btn btn-info btn-xs" (click)="enroll(course.id)"> Enroll </button>
```
Comment (wkurniawan07) [inline src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.html]: Do you really need this (`enroll`) to be a function? Judging by the usage you can just pass `routerLink` to this button element.

File: src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts [hunk 12]
```diff
@@ -12,13 +69,494 @@ import { ActivatedRoute } from '@angular/router';
...
+    const paramMap: { [key: string]: string } = { user };
+    this.httpRequestService.get('/instructor/students/courses', paramMap)
+        .subscribe((resp: GetCourseResponse) => {
+          this.courses = resp.courses;
+
+          if (!this.courses) {
+            this.statusMessageService.showWarningMessage('You do not have any courses yet.');
+          }
+
+          if (this.isDisplayArchive === undefined) {
```
Comment (wkurniawan07) [inline src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts]: When will this condition ever be achieved? `this.isDisplayArchive` is already initialized as `false`.

File: src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts [hunk 13]
```diff
@@ -12,13 +69,494 @@ import { ActivatedRoute } from '@angular/router';
...
+            this.courseStudentListSectionDataMap[courseDetails.id]
+                = this.getStudentListSectionDataFromCourseDetails(courseDetails);
+          }
+        }, (error: any) => {
+          this.statusMessageService.showErrorMessage(error);
+        });
+  }
+
+  /**
+   * Function to call backend API to get CourseDetails data for a course.
```
Comment (wkurniawan07) [inline src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts]: - ~~Function to~~
- ~~Call backend API~~. It doesn't matter whether it needs to call back end API or whatever means it needs to be done; the point is this function aims to get the course details and that's all that is necessary.

File: src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts [hunk 14]
```diff
@@ -12,13 +69,494 @@ import { ActivatedRoute } from '@angular/router';
...
+            return this.getCourseDetails(resp.course.id);
+          }
+
+          this.statusMessageService.showErrorMessage(`Error retrieving course details for course of id ${courseid}`);
+          return null;
+        }));
+  }
+
+  /**
+   * If the courseDetails of a course is fetched.
```
Comment (wkurniawan07) [inline src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts]: - You seem to miss `Returns true` at the beginning.
- "is fetched" and "in list" doesn't seem to tally.

File: src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts [hunk 15]
```diff
@@ -12,13 +69,494 @@ import { ActivatedRoute } from '@angular/router';
...
+          }
+        }, (error: any) => {
+          this.statusMessageService.showErrorMessage(error);
+        });
+  }
+
+  /**
+   * Function to call backend API to get CourseDetails data for a course.
+   */
+  fetchCourseDetails(courseid: string): Observable<CourseDetails | null> {
```
Comment (wkurniawan07) [inline src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts]: Do you really need a return value for this function?

File: src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts [hunk 16]
```diff
@@ -12,13 +69,494 @@ import { ActivatedRoute } from '@angular/router';
...
+      this.studentDetailsList.push(student);
+    }
+  }
+
+  /*------------------------------Functions to get item state from state maps-----------------------------------*/
+  /**
+   * Function to get the list of students which need to be hidden on the StudentTable display.
+   */
+  getStudentsToHide(courseId: string): string[] {
+    const studentsToHide: string[] = this.studentDetailsList
```
Comment (wkurniawan07) [inline src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts]: You can just return this; no need to assign it to a variable.

File: src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts [hunk 17]
```diff
@@ -12,13 +69,494 @@ import { ActivatedRoute } from '@angular/router';
...
+  /**
+   * Function to formulate data for student list.
+   */
+  getStudentListSectionDataForCourse(courseId: string): StudentListSectionData[] {
+    const data: StudentListSectionData[] | undefined = this.courseStudentListSectionDataMap[courseId];
+    if (data === undefined) {
+      return [];
+    }
+
+    return data;
```
Comment (wkurniawan07) [inline src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts]: One-liner alternative:
```ts
return this.courseStudentListSectionDataMap[courseId] || [];
```

File: src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts [hunk 18]
```diff
@@ -12,13 +69,494 @@ import { ActivatedRoute } from '@angular/router';
...
+    const students: StudentListStudentData[] = [];
+
+    sectionDetails.teams.forEach((team: TeamDetails) => {
+      team.students.forEach((student: StudentDetails) => {
+        const studentData: StudentListStudentData = this.mapStudentForStudentList(student);
+        students.push(studentData);
+      });
+    });
+
+    const section: StudentListSectionData = {
```
Comment (wkurniawan07) [inline src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts]: You could just return this directly. If the typing gives you problem, you should mention `as StudentListSectionData`.

File: src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts [hunk 19]
```diff
@@ -12,13 +69,494 @@ import { ActivatedRoute } from '@angular/router';
...
+    };
+
+    return section;
+  }
+
+  /**
+   * Function to formulate data for student list from StudentDetails.
+   */
+  mapStudentForStudentList(studentDetails: StudentDetails): StudentListStudentData {
+    const student: StudentListStudentData = {
```
Comment (wkurniawan07) [inline src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts]: Similarly here.

File: src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts (hunk 6 above)
Comment (wkurniawan07) [reply src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts]: Will a little help of `!!` work?
```ts
return !!this.allPresentCourses.length && this.allPresentCourses.length === this.allCheckedCourses.length;
```

File: src/main/java/teammates/common/datatransfer/DataBundle.java [hunk 20]
```diff
@@ -45,6 +48,57 @@ public void sanitizeForSaving() {
...
+
+    public List<StudentAttributes> getStudentsForTeam(String courseId, String sectionName, String teamName) {
+        return this.students.values()
+                .stream()
+                .filter(student -> student.course.equals(courseId))
+                .filter(student -> student.section.equals(sectionName))
+                .filter(student -> student.team.equals(teamName))
+                .collect(Collectors.toList());
+    }
+
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/DataBundle.java]: All of the above methods are used only in test. Let's not add them in production class.

If my observation is correct, you can move all of the above methods to the sole test that uses them.

File: src/main/java/teammates/ui/controller/InstructorSearchPageAction.java [hunk 21]
```diff
@@ -1,68 +0,0 @@
-package teammates.ui.controller;
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/controller/InstructorSearchPageAction.java]: I don't think search-related stuffs are part of this PR...

Comment (wenmogu) [reply src/main/java/teammates/ui/controller/InstructorSearchPageAction.java]: yup u r right...thats how i got myself into this conflict...
but search page was already implemented? just wondering y it is still not deleted

Comment (wkurniawan07) [reply src/main/java/teammates/ui/controller/InstructorSearchPageAction.java]: Not entirely; searching for feedback response comments are still not migrated yet.

Comment (wenmogu) [reply src/main/java/teammates/ui/controller/InstructorSearchPageAction.java]: oh i see
wokie i will roll back then ;_;

File: src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts (hunk 15 above)
Comment (wenmogu) [reply src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts]: yes i do. Because this function is to retrieve course data from backend, and hence we need to perform a check whether to carry out such retrieval. Processing of the retrieved data should be outside this function (see [line 276](https://github.com/TEAMMATES/teammates/pull/9374/files#diff-3e7add916ff8b5693f7e8cb03b522b9aR276))
